Add messages for force result download Concent use case

rwrzesien opened the pull request:
https://github.com/golemfactory/golem-messages/issues/76

lukasz-glen commented on golem_messages/message/concents.py:
```diff
@@ -147,6 +147,67 @@ class FileInfo(datastructures.FrozenDict):
+    TYPE = CONCENT_MSG_BASE + 6
+
+    __slots__ = [
+        'report_computed_task',
```
According to the documentation (Dokumentacja analityczna integracji Golem-Concent) a field is missing: there should be also an optional MessageForceReportComputedTask.

lukasz-glen commented on golem_messages/message/concents.py:
```diff
@@ -147,6 +147,67 @@ class FileInfo(datastructures.FrozenDict):
+
+class ForceGetTaskResultAck(base.Message):
+    TYPE = CONCENT_MSG_BASE + 7
+
```
According to the documentation (Dokumentacja analityczna integracji Golem-Concent) a field is missing: there should be also a mandatory MessageForceGetTaskResult.

lukasz-glen commented on golem_messages/message/concents.py:
```diff
@@ -147,6 +147,67 @@ class FileInfo(datastructures.FrozenDict):
+    TYPE = CONCENT_MSG_BASE + 9
+
+    __slots__ = [
+        'reason',
```
According to the documentation (Dokumentacja analityczna integracji Golem-Concent) a field is missing: there should be also a mandatory MessageForceGetTaskResult.

jiivan commented on golem_messages/message/__init__.py:
```diff
@@ -54,6 +54,11 @@
+from golem_messages.message.concents import ForceGetTaskResultAck  # noqa
+from golem_messages.message.concents import ForceGetTaskResultFailed  # noqa
+from golem_messages.message.concents import ForceGetTaskResultRejected  # noqa
+from golem_messages.message.concents import ForceGetTaskResultUpload  # noqa
```
Please don't import these messages here. Previous messages are imported for backwards compatibility only.

jiivan commented on golem_messages/message/__init__.py:
```diff
@@ -136,7 +141,12 @@ def init_messages():
+            ForceGetTaskResultAck,
+            ForceGetTaskResultFailed,
+            ForceGetTaskResultRejected,
+            ForceGetTaskResultUpload, ):
```
Please use `concents` namespace.

rwrzesien commented:
@jiivan @lukasz-glen I know that pylint check failed for this, but actually it is complaining that this part of code is repeated:
```
    ENUM_SLOTS = {
        'reason': REASON,
    }

    def deserialize_slot(self, key, value):
        value = super().deserialize_slot(key, value)
```
This is something I can fix only by switching them upside down which will decrease readability. Can I add pylint-ignore for this ?

jiivan commented:
@rwrzesien how about this?

```diff
diff --git i/golem_messages/message/base.py w/golem_messages/message/base.py
index 40a067d..2fcc642 100644
--- i/golem_messages/message/base.py
+++ w/golem_messages/message/base.py
@@ -306,6 +306,17 @@ class Message():
         return (name not in Message.__slots__) and (name in self.__slots__)
 
 
+class AbstractReasonMessage(Message):
+    __slots__ = [
+        'reason',
+    ] + Message.__slots__
+
+    @property
+    def ENUM_SLOTS(self):
+        return {
+            'reason': self.REASON,
+        }
+
 ##################
 # Basic Messages #
 ##################
@@ -345,11 +356,11 @@ class RandVal(Message):
     __slots__ = ['rand_val'] + Message.__slots__
 
 
-class Disconnect(Message):
+class Disconnect(AbstractReasonMessage):
     TYPE = 2
     ENCRYPT = False
 
-    __slots__ = ['reason'] + Message.__slots__
+    __slots__ = AbstractReasonMessage.__slots__
 
     class REASON(enum.Enum):
         DuplicatePeers = 'duplicate_peers'
@@ -365,10 +376,6 @@ class Disconnect(Message):
         KeyNotDifficult = 'key_not_difficult'
         Bootstrap = 'bootstrap'
 
-    ENUM_SLOTS = {
-        'reason': REASON,
-    }
-
 
 class ChallengeSolution(Message):
     TYPE = 3
diff --git i/golem_messages/message/concents.py w/golem_messages/message/concents.py
index ae978db..1b08ab5 100644
--- i/golem_messages/message/concents.py
+++ w/golem_messages/message/concents.py
@@ -186,22 +186,17 @@ class ForceGetTaskResultFailed(base.Message):
         return tasks.deserialize_task_to_compute(key, value)
 
 
-class ForceGetTaskResultRejected(base.Message):
+class ForceGetTaskResultRejected(base.AbstractReasonMessage):
     TYPE = CONCENT_MSG_BASE + 9
 
     __slots__ = [
-        'reason',
         'force_get_task_result',
-    ] + base.Message.__slots__
+    ] + base.AbstractReasonMessage.__slots__
 
     class REASON(enum.Enum):
         OperationAlreadyInitiated = 'operation_already_initiated'
         AcceptanceTimeLimitExceeded = 'acceptance_time_limit_exceeded'
 
-    ENUM_SLOTS = {
-        'reason': REASON,
-    }
-
     def deserialize_slot(self, key, value):
         value = super().deserialize_slot(key, value)
         return deserialize_force_get_task_result(key, value)
diff --git i/golem_messages/message/tasks.py w/golem_messages/message/tasks.py
index 054d7e5..de2b835 100644
--- i/golem_messages/message/tasks.py
+++ w/golem_messages/message/tasks.py
@@ -76,22 +76,17 @@ class TaskToCompute(base.Message):
         return value
 
 
-class CannotAssignTask(base.Message):
+class CannotAssignTask(base.AbstractReasonMessage):
     TYPE = TASK_MSG_BASE + 3
 
     __slots__ = [
-        'reason',
         'task_id'
-    ] + base.Message.__slots__
+    ] + base.AbstractReasonMessage.__slots__
 
     class REASON(enum.Enum):
         NotMyTask = 'not_my_task'
         NoMoreSubtasks = 'no_more_subtasks'
 
-    ENUM_SLOTS = {
-        'reason': REASON,
-    }
-
 
 class ReportComputedTask(base.Message):
     # FIXME this message should be simpler
@@ -227,14 +222,13 @@ class WaitingForResults(base.Message):
     __slots__ = base.Message.__slots__
 
 
-class CannotComputeTask(base.Message):
+class CannotComputeTask(base.AbstractReasonMessage):
     TYPE = TASK_MSG_BASE + 26
 
     __slots__ = [
-        'reason',
         'subtask_id',
         'task_to_compute',
-    ] + base.Message.__slots__
+    ] + base.AbstractReasonMessage.__slots__
 
     class REASON(enum.Enum):
         WrongCTD = 'wrong_ctd'
@@ -244,10 +238,6 @@ class CannotComputeTask(base.Message):
         NoSourceCode = 'no_source_code'
         WrongDockerImages = 'wrong_docker_images'
 
-    ENUM_SLOTS = {
-        'reason': REASON,
-    }
-
     def deserialize_slot(self, key, value):
         value = super().deserialize_slot(key, value)
         return deserialize_task_to_compute(key, value)
```

rwrzesien commented:
@jiivan Looks good, I will add it. Don't you want this class to be called ReasonMessageMixin instead ?

jiivan commented:
I would call it MixIn if it inherited from `object` not from `base.Message`. It's more abstract IMHO.

rwrzesien commented:
@shadeofblue I don't know, is it? How I can check it ?

shadeofblue commented:
@rwrzesien well - I think each `Message` class should have at least some tests that generally check if it's functioning correctly. And additional tests to verify correct functioning if there are different scenarios in the specific message's usage.

@jiivan if you don't think it's critical for each `Message` to be test-covered, please just dismiss my review. Otherwise, I'd feel safer to think we have at least minimal test coverage here...

shadeofblue commented on golem_messages/message/concents.py:
```diff
@@ -147,6 +147,76 @@ class FileInfo(datastructures.FrozenDict):
+
+
+class ForceGetTaskResultAck(base.Message):
+    TYPE = CONCENT_MSG_BASE + 7
```
@rwrzesien ah, and also, please update these `TYPE`s so that they don't collide with the other batch of newly-added messages that I have merged today ... 

(yeah, I know it's pain - I guess we could re-think the way these are assigned)

rwrzesien commented:
@shadeofblue 
> I think each Message class should have at least some tests that generally check if it's functioning correctly.

I think there are, at least at first run of `setup.py test` I have some issues I had to fix.

rwrzesien commented:
@shadeofblue @jiivan @lukasz-glen updated.

shadeofblue commented:
@rwrzesien hmm - have you specifically added any tests for those message classes that you have added? if not, imvho, you should have... so of course, there are some test that might have been broken and it's good that you have fixed them but if you have added any new message classes, you should have added tests for those specific classes similarly how there already are tests for each of the already existing message classes...

rwrzesien commented:
@shadeofblue at first I thought that there are already some tests which automatically tests every new message, because when I have added new ones I get some errors in tests for those new messages. and I had to fix them (messages code, not tests) but I have took a look and I have seen that there aren't this kind of tests, so I have added tests specific for new messages.

rwrzesien commented:
@jiivan As @lukasz-glen and @shadeofblue are out of office this week, could you please review again, taking their comments and my updates into consideration, and accept or ask for new changes ? We need this quite soon to test use case we need to provide soon.
